File: DSA/src/Questions/Array/RectangleIntersections.java
package Questions.Array;

import java.util.ArrayList;
import java.util.List;

public class RectangleIntersections {
    static class Rectangle {
        int[] bottomLeft;
        int[] topRight;

        Rectangle(int[] bottomLeft, int[] topRight) {
            this.bottomLeft = bottomLeft;
            this.topRight = topRight;
        }
    }

    public static int countIntersectionPoints(int[] linesX, int[] linesY, List<Rectangle> rectangles) {
        int intersectionCount = 0;

        // Iterate through lines parallel to X-axis
        for (int lineX : linesX) {
            for (Rectangle rectangle : rectangles) {
                if (rectangle.bottomLeft[1] <= lineX && lineX <= rectangle.topRight[1]) {
                    intersectionCount++;
                }
            }
        }

        // Iterate through lines parallel to Y-axis
        for (int lineY : linesY) {
            for (Rectangle rectangle : rectangles) {
                if (rectangle.bottomLeft[0] <= lineY && lineY <= rectangle.topRight[0]) {
                    intersectionCount++;
                }
            }
        }

        return intersectionCount;
    }

    public static void main(String[] args) {
        int[] linesX = {1, 2, 4, 19};
        int[] linesY = {2, 5, 9};
        List<Rectangle> rectangles = new ArrayList<>();
        rectangles.add(new Rectangle(new int[]{1, 3}, new int[]{3, 6}));

        // Count intersection points
        int totalIntersectionPoints = countIntersectionPoints(linesX, linesY, rectangles);
        System.out.println("Total intersection points: " + totalIntersectionPoints);
    }

}

File: DSA/Python/Questions/Array/RectangleIntersections.py
"""
# Question: Rectangle Intersections
# Count intersection points between rectangles and grid lines

# Time Complexity: O(n * (lx + ly)) where n is number of rectangles, 
# lx and ly are number of lines
# Space Complexity: O(1)

# Algorithm:
# 1. Define Rectangle class to store coordinates
# 2. Check intersections with X-axis parallel lines
# 3. Check intersections with Y-axis parallel lines
# 4. Count total intersections

# Key Components:
# - Rectangle class for coordinate storage
# - count_intersection_points(): Main counting function
# - Separate X and Y axis intersection checks
"""

from typing import List, Tuple
from dataclasses import dataclass


@dataclass
class Rectangle:
    bottom_left: Tuple[int, int]
    top_right: Tuple[int, int]


class RectangleIntersections:
    def count_intersection_points(
        self, lines_x: List[int], lines_y: List[int], rectangles: List[Rectangle]
    ) -> int:
        intersection_count = 0

        # Check X-axis parallel lines
        for line_x in lines_x:
            for rectangle in rectangles:
                if rectangle.bottom_left[1] <= line_x <= rectangle.top_right[1]:
                    intersection_count += 1

        # Check Y-axis parallel lines
        for line_y in lines_y:
            for rectangle in rectangles:
                if rectangle.bottom_left[0] <= line_y <= rectangle.top_right[0]:
                    intersection_count += 1

        return intersection_count


def main():
    lines_x = [1, 2, 4, 19]
    lines_y = [2, 5, 9]
    rectangles = [Rectangle((1, 3), (3, 6))]

    solution = RectangleIntersections()
    result = solution.count_intersection_points(lines_x, lines_y, rectangles)
    print(f"Total intersection points: {result}")


if __name__ == "__main__":
    main()
